### PR TITLE
Launchpad: Migrate start-writing flow to use the new structure

### PR DESCRIPTION
**Pull request by gabrielcaires**
Related to #86647

## Proposed Changes
Migrate the tasks which are part of the flow `start-writing` 

| TASK | MIGRATED? |
|--------|--------|
| first_post_published |  Previously Migrated |
| setup_blog | THIS PR |
| 'domain_upsell' | Previously Migrated |
| plan_completed | THIS PR |
| blog_launched | THIS PR | 

### Flags Expected State
**production**:                    false
**development**:                   true
**stage**:                         true
**test**:                          true

## Testing Instructions
* Start the flow `start-writing`  (setup/start-writing)
* Choose 'New site' 
* Set a post title and content
* Publish the post
* Wait the flow redirect you to `setup/start-writing/launchpad?siteSlug=[YOUR_SITE]`
* Enable the feature flags via URL and reload
* Check if  there are messages on devtools to all task above
* Use the launchpad (Set values, Edit values, Check if the values are marked when completed)
* Finish the flow.

Requires the feature flags:
* "launchpad/new-task-definition-parser"
* "launchpad/new-task-definition-parser/start-writing"

`flags=launchpad/new-task-definition-parser,launchpad/new-task-definition-parser/start-writing` 

The flags can be enabled appending  the `?flags=` flag on the url (Please always check if the flags= param is still on the url when the system redirect back the user from some "no launchpad" page ).

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/types.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ export type { Task } from '@automattic/launchpad';
 import { ChecklistStatuses, SiteDetails } from '@automattic/data-stores';
 import { Task } from '@automattic/launchpad';
 import { MinimalRequestCartProduct } from '@automattic/shopping-cart';
+import { ReactNode } from 'react';
```
Comment (daledupreez) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/types.ts]: Nit: this should be a type import.
```suggestion
import { type ReactNode } from 'react';
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx [hunk 2]
```diff
@@ -14,6 +15,19 @@ const getSetupFreeTask: TaskAction = ( task, flow, context ): Task => {
 	};
 };
 
+const getSetupBlog: TaskAction = ( task, flow, context ): Task => {
+	const { siteInfoQueryArgs } = context;
+
+	return {
+		...task,
+		actionDispatch: () => recordTaskClickTracksEvent( task, flow, context ),
```
Comment (daledupreez) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx:23]: I have a concern around the fact that we're implementing what should be generic Tracks triggers in this task-specific logic. This should be somewhere central so we don't have cases where it's accidentally left off for some tasks, or where we make some change that doesn't get rolled out to all locations. Can this be moved into `getTaskDefinition()`?

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import { Task } from '@automattic/launchpad';
+import { isBlogOnboardingFlow } from '@automattic/onboarding';
 import { addQueryArgs } from '@wordpress/url';
 import { recordTaskClickTracksEvent } from '../../tracking';
 import { TaskAction } from '../../types';
```
Comment (daledupreez) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx]: Nit: I think this should be a type import.

**Review by daledupreez — COMMENTED** [review]
I didn't get a chance to test, but I do have some feedback.

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx (hunk 2 above)
Comment (gabrielcaires) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx:23]: I agree but I prefer to finish this migration first and then iterate around it. 
My thought is avoid to introduce more changes as part of this migration can make the review harder. I will a task to do it in a separated PR

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx (hunk 3 above)
Comment (gabrielcaires) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx:5]: Fixed https://github.com/Automattic/wp-calypso/pull/86584/commits/ed8db385028cd43eeb3a4d39369dee1b0c9d1c20

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/types.ts (hunk 1 above)
Comment (gabrielcaires) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/types.ts]: Fixed https://github.com/Automattic/wp-calypso/pull/86584/commits/ed8db385028cd43eeb3a4d39369dee1b0c9d1c20

**Comment by markbiek** [comment]
I wasn't able to get this to work.

First, I went to https://container-optimistic-johnson.calypso.live/setup/start-writing?flags=launchpad/new-task-definition-parser/start-writing,launchpad/new-task-definition-parser.

It asked me to choose a new or existing site and I chose a new site. Then it took me directly to the post editor.

On the second try, when I got to the "choose a new or existing site" screen, I tried manually changing the url to https://container-optimistic-johnson.calypso.live/setup/start-writing/launchpad?flags=launchpad/new-task-definition-parser/start-writing,launchpad/new-task-definition-parser.

In that case, the launchpad flashed for a second and I was redirected to `/sites`.

I didn't see any messages about task definitions in the dev console.

https://github.com/Automattic/wp-calypso/assets/917632/5836cad7-bab8-4436-98b7-e8efea0aacf0

**Comment by gabrielcaires** [comment]
> I wasn't able to get this to work.

@markbiek Sorry my fault, this flow Is not intuitive and I didn't mention it.
The issue we are facing seems to be related to the usability of the flow. The primary concern is at the 0:20 mark in your video, where you need to enter the title post content and publish the post. After that, the application redirects you to the appropriate launchpad screen. On the launchpad screen, you need to set the flag on the URL and only at this moment, the launchpad should use the new task definitions.

**Comment by markbiek** [comment]
> > I wasn't able to get this to work.
> 
> @markbiek Sorry my fault, this flow Is not intuitive and I didn't mention it. The issue we are facing seems to be related to the usability of the flow. The primary concern is at the 0:20 mark in your video, where you need to enter the title post content and publish the post. After that, the application redirects you to the appropriate launchpad screen. On the launchpad screen, you need to set the flag on the URL and only at this moment, the launchpad should use the new task definitions.

Thanks for updating the testing instructions! This is now working as expected.

Does any of @daledupreez 's feedback still need to be done or can this be approved?

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx (hunk 3 above)
Comment (gabrielcaires) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-definitions/setup/index.tsx]: Fixed here https://github.com/Automattic/wp-calypso/pull/86584/commits/511e0cebb0f7e9d536dafc34e4494dec6d5f251f

**Comment by gabrielcaires** [comment]
@markbiek I think we can go, missing feedback I will address as part of the code cleanup after merge all changes

**Review by markbiek — APPROVED** [review]
LGTM!

**Pull request merged**